libeos-payg/service.c

```diff
@@ -125,22 +126,90 @@ epg_service_get_main_option_entries (GssService *service)
+                              gpointer      user_data);
 static void manager_new_cb (GObject      *source_object,
                             GAsyncResult *result,
                             gpointer      user_data);
+static void epg_service_startup_complete (EpgService  *self,
+                                          GTask       *task,
+                                          EpgProvider *provider);
 
 static void
 epg_service_startup_async (GssService          *service,
                            GCancellable        *cancellable,
                            GAsyncReadyCallback  callback,
                            gpointer             user_data)
 {
-  EpgService *self = EPG_SERVICE (service);
   g_autoptr(GError) local_error = NULL;
 
   g_autoptr(GTask) task = g_task_new (service, cancellable, callback, user_data);
   g_task_set_source_tag (task, epg_service_startup_async);
 
+  g_autoptr(EpgProviderLoader) loader = epg_provider_loader_new (NULL);
+  epg_provider_loader_load_async (loader, cancellable, provider_load_cb,
+                                  g_steal_pointer (&task));
+}
+
+static void
+provider_load_cb (GObject      *source_object,
+                  GAsyncResult *result,
+                  gpointer      user_data)
+{
+  g_autoptr(GTask) task = G_TASK (user_data);
+  EpgService *self = EPG_SERVICE (g_task_get_source_object (task));
+  EpgProviderLoader *loader = EPG_PROVIDER_LOADER (source_object);
+  g_autoptr(GPtrArray) providers = NULL;
+  g_autoptr(GError) local_error = NULL;
+
+  providers = epg_provider_loader_load_finish (loader, result, &local_error);
+  if (providers == NULL)
+    {
+      g_warning ("%s: Failed to load external providers: %s",
+                 G_STRFUNC, local_error->message);
+      g_clear_error (&local_error);
+    }
+  else
+    {
+      EpgProvider *provider = NULL;
+      guint i;
+
+      /* Find the first enabled external provider, and use that. */
+      for (i = 0; i < providers->len; i++)
+        {
+          provider = EPG_PROVIDER (g_ptr_array_index (providers, i));
+          if (epg_provider_get_enabled (provider))
+            {
+              g_debug ("%s: Found enabled external provider %s",
+                       G_STRFUNC, G_OBJECT_TYPE_NAME (provider));
+              g_ptr_array_remove_index_fast (providers, i);
+              break;
+            }
+          else
+            {
+              g_debug ("%s: external provider %s is not enabled",
+                       G_STRFUNC, G_OBJECT_TYPE_NAME (provider));
+              provider = NULL;
+            }
+        }
+
+      /* Shut down all providers except the first enabled one (if any). */
+      for (i = 0; i < providers->len; i++)
+        {
+          EpgProvider *spare = EPG_PROVIDER (g_ptr_array_index (providers, i));
+          epg_provider_shutdown_async (spare, NULL, NULL, NULL);
+        }
+
+      if (provider != NULL)
+        {
+          epg_service_startup_complete (self, task, provider);
```

Comment:
I’d do `g_steal_pointer (&provider)` instead of just passing `provider` in order to make it explicit that this takes ownership of it. Otherwise this could look like a memory leak.